sheet_to_triples/tests/test_main.py

```diff
@@ -0,0 +1,267 @@
+# Copyright 2021 Visual Meaning Ltd
+# This is free software licensed as GPL-3.0-or-later - see COPYING for terms.
+
+"""Unittests for the __main__ module of sheet-to-triples."""
+
+import io
+import json
+import os
+import unittest
+
+from unittest import mock
+
+from .. import __main__ as main, run
+
+
+class StubTransform:
+
+    def __init__(self, name):
+        self.name = name
+        self.triples = []
+
+    def uses_sheet(self):
+        if 'book' in self.name:
+            return True
+        return False
+
+    @classmethod
+    def iter_from_name(cls, name, base_path=''):
+        return [cls(os.path.join(base_path, name) + '.py')]
+
+
+def _mock_os_path_walk(walk_data):
+    return mock.patch('os.walk', return_value=walk_data)
+
+
+def _mock_os_path_dirname(dirname):
+    return mock.patch('os.path.dirname', return_value=dirname)
+
+
+def _mock_os_path_isdir(isdir):
+    return mock.patch('os.path.isdir', return_value=isdir)
+
+
+def _mock_open(data):
+    return mock.patch('builtins.open', mock.mock_open(read_data=data))
+
+
+def _mock_stderr(buffer):
+    return mock.patch('sys.stderr', new=buffer)
+
+
+_iter_from_name = 'sheet_to_triples.trans.Transform'
+
+
+@mock.patch(_iter_from_name, new=StubTransform('dummy'))
+class TestParseArgs(unittest.TestCase):
+
+    def test_parse_args_default(self):
+        argv = ['dummy', 'transform1']
+        args = main.parse_args(argv)
+        defaults = [
+            ('book', None),
+            ('add_graph', None),
+            ('model', None),
+            ('model_out', 'new.json'),
+            ('resolve_same', True),
+            ('debug', False),
+            ('verbose', False),
+            ('from_list', None),
+            ('non_unique_from', None),
+        ]
+
+        for arg, expected in defaults:
+            with self.subTest(arg=arg):
+                self.assertEqual(getattr(args, arg), expected)
+
+    def test_parse_args_transform(self):
+        argv = ['dummy', 'transform1', 'transform2']
+        args = main.parse_args(argv)
+
+        self.assertEqual(
+            [t.name for t in args.transform],
+            ['transform1.py', 'transform2.py']
+        )
+
+    def test_parse_args_transform_error_if_requires_book(self):
+        argv = ['dummy', 'booktransform1']
+        buffer = io.StringIO()
+
+        # this makes any unexpected lower level exceptions harder to debug
+        # TODO: come up with a nicer way of suppressing "normal" error output
+        with _mock_stderr(buffer), self.assertRaises(SystemExit) as e:
```

Comment:
Yeah, this is awkward. Can also do it by patching out [`ArgumentParser.exit`](https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.exit) which is a bit less mess.